Haena : 카드클래스 구현 (3/14)

boriiiborii opened the pull request:
makeCard에 숫자, 모양, 이모티콘 외에 이미지를 넣은 이유는 
실행시켰을때 화면에 2-1에서의 카드 뒷면 이미지를 해당하는 이미지로 띄우고 확인하기 위해 사용

shape에서 enum은 4가지 모양밖에 없어서 사용하였으며
이모티콘과 이미지는 모양에 따라 정해지기 때문에 스위치로 구현

+함수 내에 printCard메서드 추가로 구현하였습니다.
(프로그래밍 요구사항에 카드 클래스를 구현하는 곳에 메서드를 구현하라고 적혀있어서 수정)

혹시나 틀린게 있다면 커맨트 부탁드립니다.

boriiiborii committed:
2-2 체크리스트 작성

boriiiborii committed:
step2-1 코드리뷰 반영

프로젝트 생성 이후에 만들어지는 코드 중 필요없는 코드 삭제
배경이미지가 없더라도 앱이 꺼지지 않고 작업을 할 수 있도록 개선
의미가 있는 숫자들 변수처리

boriiiborii committed:
객체지향 프로그래밍 방식에 충실하게 카드 클래스 설계

MakeCard클래스에 설계하였음

boriiiborii committed:
카드 정보를 출력하기 위한 문자열을 반환하는 함수를 구현

var haveCard = Array(repeating: MakeCard(shape: .clover, number: 1), count: 7)
이렇게 카드배열에 클로버1을 7번씩 다 채워넣고
func printFirstCardInformation() {
        print("\((haveCard.first?.imoticon)!) \((haveCard.first?.number)!)")
    }
코드를 사용하여 첫번째 배열에 있는 문자열을 반환하였다.

boriiiborii committed:
ViewController에서 특정한 카드 객체 인스턴스를 만들어서 콘솔에 출력

리드미 하단에 사진 첨부하였음

godrm commented:
PR 메시지도 가이드에 맞게 좀 더 풍성하게 작성해보세요.

godrm commented on PokerGame/PokerGame/MakeCard.swift:
```diff
@@ -0,0 +1,63 @@
+import Foundation
+import UIKit
+
+class MakeCard{
```
MakeCard가 동사형 표현이 섞여 있어서 클래스 이름으로 어색하네요
그냥 Card가 아니라 MakeCard인 이유가 있나요?

boriiiborii replied:
카드를 하나씩 인스턴스로 생성한다고 makeCard라는 이름을 사용했는데 make를 빼는게 더 자연스러울거 같네요 수정하겠습니다.

boriiiborii committed:
makeCard 클래스 내부에 printCard함수 생성

godrm commented on PokerGame/PokerGame/MakeCard.swift:
```diff
@@ -0,0 +1,63 @@
+    //shape enum타입 한 이유 : 어짜피 모양은 4개중 하나니까
+    var shape : Shape
+    //number을 String으로 한 이유: AJQK로 변환을 해야하기 때문에 숫자로 받고 String으로 저장하는 형식으로 진행
+    var number : String
```
Shape는 Enum으로 선언했는데 number는 String이네요. 
number도 Enum으로 동일한 수준으로 맞춰서 선언해보실래요? 
타입을 선언해서 어떤 역할인지 더 명확하게 만들 수 있습니다.

boriiiborii replied:
넵 수정하겠습니다.

godrm commented on PokerGame/PokerGame/MakeCard.swift:
```diff
@@ -0,0 +1,63 @@
+    //number을 String으로 한 이유: AJQK로 변환을 해야하기 때문에 숫자로 받고 String으로 저장하는 형식으로 진행
+    var number : String
+    var imoticon : String
+    var image : UIImage
```
여기 Card는 모델 역할을 하는 객체라서 UIKit에 의존하지 않는 게 좋습니다. 
UIImage를 여기에 보관할 필요가 없고, 카드 인스턴스에 따라서 카드 이미지와 매칭하는 게 더 좋습니다. 
모든 것을 다 갖는 타입을 선언해야 하는 것은 아닙니다

boriiiborii replied:
카드 인스턴스를 찍어낼때 shape에 따라서 이미지를 인스턴스에 바로 저장할 수 있도록 하고싶었는데, 
shape와 이미지가 연관이 있으니 인스턴스를 생성하고 후에 이미지와 매칭을 시키는것이 좋다는 말씀인건가요 ??

godrm replied:
shape에 따라서 이미지를 매칭한다는 것 자체가 ViewController에서 고민이잖아요? 그건 거기서 구현해야 한다는거죠
여기에서는 UIKit을 import할 이유가 전혀 없습니다

godrm commented on PokerGame/PokerGame/MakeCard.swift:
```diff
@@ -0,0 +1,63 @@
+    var imoticon : String
+    var image : UIImage
+    
+    init(shape: Shape, number: Int) {
+        
+        self.shape = shape
```
카드 타입을 생성하는 관점에서도 Shape와 Int 가 레벨/수준이 맞지 않는 것 같습니다. 위에 언급한 것처럼 Enum으로 바꾸면 좀 더 명확해보일꺼 같네요

boriiiborii replied:
넵 수정하겠습니다

godrm commented on PokerGame/PokerGame/MakeCard.swift:
```diff
@@ -0,0 +1,63 @@
+    var shape : Shape
+    //number을 String으로 한 이유: AJQK로 변환을 해야하기 때문에 숫자로 받고 String으로 저장하는 형식으로 진행
+    var number : String
+    var imoticon : String
```
imoticon 이라는 문자는 Shape에 따라서 결정되니까 두 타입이 서로 관련이 있는 것 같습니다. 
어느 하나는 없어도 될꺼 같네요!

boriiiborii replied:
넵 !!!

godrm commented on PokerGame/PokerGame/MakeCard.swift:
```diff
@@ -0,0 +1,63 @@
+        case .diamond :
+            self.imoticon = "◆"
+            self.image = UIImage(named: "diamond")!
+        }
```
init이 모든 로직을 다 갖고 있는 구조가 되네요
여기 코드를 보면 number에 대한 처리, shape에 대한 처리로 나눌 수 있습니다
이 부분이 꼭 init에서 다 처리할 필요가 없고, number 타입이나 shape 타입에서 처리할 수도 있습니다
하위 타입을 활용해보세요

godrm commented on PokerGame/PokerGame/ViewController.swift:
```diff
@@ -9,26 +9,41 @@ import UIKit
+        printFirstCardInformation()
+    }
+    
+    func printCardImage() {
```
여기 구현부분은 카드 뷰를 생성해서 추가하는 거라 printCardImage 카드 이미지를 print하는 동작이 아니지 않나요? 
함수 이름이 동작을 잘 설명하지 못하네요

boriiiborii replied:
프린트이모티콘이 적절한것같네요 수정하겠습니다

godrm commented on PokerGame/PokerGame/ViewController.swift:
```diff
@@ -9,26 +9,41 @@ import UIKit
+        let cardWidth = (self.view.frame.width / numberCards)-spacingCard
+        let cardHeight = cardWidth * ratio
+        for index in 0..<haveCard.count {
+            let cardImageView = UIImageView(image: haveCard[index].image)
```
이 부분이 카드에서 설명한 부분인데 이미지를 처리하는 것은 View나 ViewController에서 다룰 부분이고, 
Model 역할을 하는 카드 타입에서는 UIImage를 갖고 있을 필요는 없습니다. 이 부분 개선해보세요

godrm commented on PokerGame/PokerGame/ViewController.swift:
```diff
@@ -9,26 +9,41 @@ import UIKit
     }
+    
+    func printFirstCardInformation() {
+        print("\((haveCard.first?.imoticon)!) \((haveCard.first?.number)!)")
```
옵셔널 타입에 대해서 강제 언래핑은 권장하지 않습니다. 
항상 `!`는 사용하지 말아야지. 생각하고 안전하게 처리하세요

boriiiborii replied:
옵셔널바인딩으로 수정하겠습니다.

godrm commented on PokerGame/PokerGame/MakeCard.swift:
```diff
@@ -0,0 +1,63 @@
+        case heart
+        case spade
+        case clover
+        case diamond
```
Shape 타입을 선언했으니까 위에 출력이나 imoticon 문자열을 매칭하는 것도 여기서 처리할 수 있을겁니다

boriiiborii replied:
넵 고려해보겠습니다

godrm reviewed (changes requested):
다음 내용들을 확인해보세요

godrm commented on PokerGame/PokerGame/MakeCard.swift:
```diff
@@ -58,6 +58,11 @@ class MakeCard{
+
+    func printCard() {
+        print(self.imoticon, self.number)
+    }
```
이 부분이 리뷰하고 나서 추가됐는데 코드에 print 구문이 있는 것은 나쁜 코드입니다. 
모델은 값을 전달해주고 출력이나 값을 확인하는 것은 더 상위에서 처리하는 게 좋습니다

boriiiborii replied:
수정하겠습니다.

boriiiborii committed:
step2-2 코드리뷰 반영

makeCard 클래스 >> Card 클래스로 수정
card클래스의 number변수 enum타입으로 지정
card클래스의 이미지 변수 삭제
card클래스의 이모티콘은 shap변수의 이넘타입(로우밸류)를 통하여 선언
card메서드 description 구현. >> print형식이 아닌 return형식으로 빼주었으며 print는 viewDidLoad에서 실행되도록 작성
이미지는 추후 미션에 있으므로 다시 카드 뒷면으로 돌려놓음
